File: insertion_sort/ruby/insertion_sort.rb
require 'byebug'

class InsertionSort

  def sort(data)
    
    sorted_data = []

    until data.empty?
      
      element_to_insert = data.shift

      inserted = false

      index = sorted_data.length - 1

      loop do

        # If this is the first element to insert, the sorted_data array is empty and we can just put the element in it. No comparison is needed.
        if sorted_data.length == 0
          sorted_data[0] = element_to_insert
          break
        elsif element_to_insert >= sorted_data.last
          sorted_data.push(element_to_insert)
          break
        elsif element_to_insert >= sorted_data[index] && element_to_insert < sorted_data[index + 1]
          sorted_data.insert(index + 1, element_to_insert)
          break
        elsif index == 0
          sorted_data.unshift(element_to_insert)
          break
        end

        index -= 1

      end

    end

  return sorted_data

  end

end

sorter = InsertionSort.new

puts sorter.sort(["d", "b", "a", "c"])

File: insertion_sort/js/insertion_sort.js
window.onload = function insertionSort() {

  var data = [];
  for (i = 0; i < 7; i++) {
    data[i] = Math.floor(Math.random() * 100);
  }

  // Put the unsorted data in the DOM
  for (i = 0; i < data.length; i++) {

    var elementContainer = document.createElement("div");
    elementContainer.className = "element-container";

    var element = document.createElement("div");
    element.innerHTML = data[i];

    elementContainer.appendChild(element);
    document.getElementById("unsorted-elements-container").appendChild(elementContainer);

  }  

  var sorted_data = [];

  // Sort the data
  while (data.length > 0) {

    var element_to_insert = data.shift();

    var index = sorted_data.length - 1;

    // If sorted array is empty, put the element_to_insert in it.
      if (sorted_data.length == 0) {
        sorted_data.push(element_to_insert);
      }

    // If the element_to_insert belongs at the end of the sorted array, put it there.
    // If the element_to_insert belongs between two elements in the sorted array, insert it there.
    // If the element_to_insert belongs at the beginning of the sorted array, insert it there.
    while (index >= 0) {
      if (element_to_insert >= sorted_data[sorted_data.length - 1]) {
        sorted_data.push(element_to_insert);
        break;
      } else if (element_to_insert >= sorted_data[index] && element_to_insert <= sorted_data[index + 1]) {
        sorted_data.splice(index + 1, 0, element_to_insert);
        break;
      } else if (index == 0) {
        sorted_data.unshift(element_to_insert);     
        break   
      }
      index --;
    }

  }
  
  // Put the sorted data in the DOM
  for (i = 0; i < sorted_data.length; i++) {

    var elementContainer = document.createElement("div");
    elementContainer.className = "element-container";

    var element = document.createElement("div");
    element.innerHTML = sorted_data[i];

    elementContainer.appendChild(element);
    document.getElementById("sorted-elements-container").appendChild(elementContainer);

  }  

}

function randomizeData() {
  location.reload();
}
